File: modules/mauth-authenticator/src/test/scala/com/mdsol/mauth/scaladsl/utils/RequestAuthenticatorTest.scala
package com.mdsol.mauth.scaladsl.utils

import java.security.Security

import com.mdsol.mauth.RequestAuthenticatorTestBase
import com.mdsol.mauth.exception.MAuthValidationException
import com.mdsol.mauth.scaladsl.RequestAuthenticator
import com.mdsol.mauth.test.utils.FakeMAuthServer.EXISTING_CLIENT_APP_UUID
import com.mdsol.mauth.util.{EpochTimeProvider, MAuthKeysHelper}
import org.bouncycastle.jce.provider.BouncyCastleProvider
import org.scalamock.scalatest.MockFactory
import org.scalatest.concurrent.ScalaFutures
import org.scalatest.{FlatSpec, Matchers}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.concurrent.duration._
import scala.language.postfixOps

class RequestAuthenticatorTest extends FlatSpec with RequestAuthenticatorTestBase with Matchers with ScalaFutures with MockFactory {

  Security.addProvider(new BouncyCastleProvider)
  private implicit val requestValidationTimeout: Duration = 10 seconds

  private val mockEpochTimeProvider = mock[EpochTimeProvider]

  "RequestAuthenticator" should "authenticate a valid request" in clientContext { (client) =>
    (mockEpochTimeProvider.inSeconds _).expects().returns(RequestAuthenticatorTestBase.CLIENT_X_MWS_TIME_HEADER_VALUE.toLong + 3)
    val authenticator = new RequestAuthenticator(client, mockEpochTimeProvider)

    whenReady(authenticator.authenticate(getSimpleRequest)) { validationResult =>
      validationResult shouldBe true
    }
  }

  it should "authenticate a request with unicode chars in body" in clientContext { (client) =>
    (mockEpochTimeProvider.inSeconds _).expects().returns(RequestAuthenticatorTestBase.CLIENT_UNICODE_X_MWS_TIME_HEADER_VALUE.toLong + 3)
    val authenticator = new RequestAuthenticator(client, mockEpochTimeProvider)

    whenReady(authenticator.authenticate(getRequestWithUnicodeCharactersInBody)) { validationResult =>
      validationResult shouldBe true
    }
  }

  it should "authenticate a request without any body" in clientContext { (client) =>
    (mockEpochTimeProvider.inSeconds _).expects().returns(RequestAuthenticatorTestBase.CLIENT_NO_BODY_X_MWS_TIME_HEADER_VALUE.toLong + 3)
    val authenticator = new RequestAuthenticator(client, mockEpochTimeProvider)

    whenReady(authenticator.authenticate(getRequestWithoutMessageBody)) { validationResult =>
      validationResult shouldBe true
    }
  }

  it should "not authenticate an invalid request" in clientContext { (client) =>
    (mockEpochTimeProvider.inSeconds _).expects().returns(RequestAuthenticatorTestBase.CLIENT_X_MWS_TIME_HEADER_VALUE.toLong + 3)
    val authenticator = new RequestAuthenticator(client, mockEpochTimeProvider)
    val invalidRequest = getSimpleRequestWithWrongSignature

    whenReady(authenticator.authenticate(invalidRequest)) { validationResult =>
      validationResult shouldBe false
    }
  }

  it should "not authenticate a request after timeout period passed" in {
    (mockEpochTimeProvider.inSeconds _).expects().returns(RequestAuthenticatorTestBase.CLIENT_UNICODE_X_MWS_TIME_HEADER_VALUE.toLong + 600)
    val authenticator = new RequestAuthenticator(mock[ClientPublicKeyProvider], mockEpochTimeProvider)

    whenReady(authenticator.authenticate(getRequestWithUnicodeCharactersInBody).failed) {
      case e: MAuthValidationException => e.getMessage shouldBe "MAuth request validation failed because of timeout 10 seconds"
      case _ => fail("should not be here")
    }
  }

  private def clientContext(test: (ClientPublicKeyProvider) => Any): Unit = {
    val client: ClientPublicKeyProvider = mock[ClientPublicKeyProvider]
    (client.getPublicKey _).expects(EXISTING_CLIENT_APP_UUID).returns(
      Future(Some(MAuthKeysHelper.getPublicKeyFromString(RequestAuthenticatorTestBase.PUBLIC_KEY))))
    test(client)
  }
}

File: modules/mauth-authenticator/src/test/java/com/mdsol/mauth/RequestAuthenticatorTest.java
package com.mdsol.mauth;

import com.mdsol.mauth.exception.MAuthValidationException;
import com.mdsol.mauth.util.MAuthKeysHelper;
import com.mdsol.mauth.utils.ClientPublicKeyProvider;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.junit.BeforeClass;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.mockito.Mockito;

import java.security.Security;

import static com.mdsol.mauth.test.utils.FakeMAuthServer.EXISTING_CLIENT_APP_UUID;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class RequestAuthenticatorTest implements RequestAuthenticatorTestBase {

  @Rule
  public ExpectedException expectedException = ExpectedException.none();

  @BeforeClass
  public static void setup() {
    Security.addProvider(new BouncyCastleProvider());
  }

  @Test
  public void validatingRequestSentAfter5MinutesShouldFail() throws Exception {
    when(mockEpochTimeProvider.inSeconds()).thenReturn(Long.parseLong(CLIENT_UNICODE_X_MWS_TIME_HEADER_VALUE) + 600);
    RequestAuthenticator authenticator = new RequestAuthenticator(mock(ClientPublicKeyProvider.class), REQUEST_VALIDATION_TIMEOUT_SECONDS, mockEpochTimeProvider);
    MAuthRequest request = getRequestWithUnicodeCharactersInBody();

    expectedException.expect(MAuthValidationException.class);
    expectedException.expectMessage("MAuth request validation failed because of timeout 300s");
    authenticator.authenticate(request);
  }

  @Test
  public void validatingInvalidRequestShouldFail() throws Exception {
    ClientPublicKeyProvider client = mock(ClientPublicKeyProvider.class);
    when(client.getPublicKey(Mockito.eq(EXISTING_CLIENT_APP_UUID))).thenReturn(MAuthKeysHelper.getPublicKeyFromString(PUBLIC_KEY));
    when(mockEpochTimeProvider.inSeconds()).thenReturn(Long.parseLong(CLIENT_X_MWS_TIME_HEADER_VALUE) + 3);
    Authenticator authenticator = new RequestAuthenticator(client, REQUEST_VALIDATION_TIMEOUT_SECONDS, mockEpochTimeProvider);
    MAuthRequest invalidRequest = getSimpleRequestWithWrongSignature();

    boolean validationResult = authenticator.authenticate(invalidRequest);

    assertThat(validationResult, equalTo(false));
  }

  @Test
  public void validatingValidRequestShouldPass() throws Exception {
    ClientPublicKeyProvider client = mock(ClientPublicKeyProvider.class);
    when(client.getPublicKey(Mockito.eq(EXISTING_CLIENT_APP_UUID))).thenReturn(MAuthKeysHelper.getPublicKeyFromString(PUBLIC_KEY));
    when(mockEpochTimeProvider.inSeconds()).thenReturn(Long.parseLong(CLIENT_X_MWS_TIME_HEADER_VALUE) + 3);
    Authenticator authenticator = new RequestAuthenticator(client, REQUEST_VALIDATION_TIMEOUT_SECONDS, mockEpochTimeProvider);
    MAuthRequest request = getSimpleRequest();

    boolean validationResult = authenticator.authenticate(request);

    assertThat(validationResult, equalTo(true));
  }

  @Test
  public void validatingValidRequestWithSpecialCharactersShouldPass() throws Exception {
    ClientPublicKeyProvider client = mock(ClientPublicKeyProvider.class);
    when(client.getPublicKey(Mockito.eq(EXISTING_CLIENT_APP_UUID))).thenReturn(MAuthKeysHelper.getPublicKeyFromString(PUBLIC_KEY));
    when(mockEpochTimeProvider.inSeconds()).thenReturn(Long.parseLong(CLIENT_UNICODE_X_MWS_TIME_HEADER_VALUE) + 3);
    Authenticator authenticator = new RequestAuthenticator(client, REQUEST_VALIDATION_TIMEOUT_SECONDS, mockEpochTimeProvider);
    MAuthRequest request = getRequestWithUnicodeCharactersInBody();

    boolean validationResult = authenticator.authenticate(request);

    assertThat(validationResult, equalTo(true));
  }

  @Test
  public void validatingValidRequestWithoutBodyShouldPass() throws Exception {
    ClientPublicKeyProvider client = mock(ClientPublicKeyProvider.class);
    when(client.getPublicKey(Mockito.eq(EXISTING_CLIENT_APP_UUID))).thenReturn(MAuthKeysHelper.getPublicKeyFromString(PUBLIC_KEY));
    when(mockEpochTimeProvider.inSeconds()).thenReturn(Long.parseLong(CLIENT_NO_BODY_X_MWS_TIME_HEADER_VALUE) + 3);
    Authenticator authenticator = new RequestAuthenticator(client, REQUEST_VALIDATION_TIMEOUT_SECONDS, mockEpochTimeProvider);
    MAuthRequest request = getRequestWithoutMessageBody();

    boolean validationResult = authenticator.authenticate(request);

    assertThat(validationResult, equalTo(true));
  }

}